src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,519 @@
+        private void PrepareContractAccount(Address recipient)
+        {
+            if (_stateProvider.AccountExists(recipient))
+            {
+                bool addressHasCode = (_virtualMachine.GetCachedCodeInfo(recipient)?.MachineCode?.Length ?? 0) != 0;
+                bool addressWasUsed = _stateProvider.GetNonce(recipient) != 0;
+                if (addressHasCode || addressWasUsed)
+                {
+                    if (_logger.IsEnabled(LogEventLevel.Verbose))
+                        _logger.Verbose($"Contract collision at {recipient}");
+                    throw new TransactionCollisionException();
+                }
+
+                _stateProvider.UpdateStorageRoot(recipient, Keccak.EmptyTreeHash);
+            }
+        }
+
+        private void InitEntryExecution(ExecutionEnvironment env, ulong gasLimit, IReleaseSpec spec, ITxTracer txTracer)
+        {
+            // first we increment nonce on the executing account
+            _stateProvider.IncrementNonce(env.Sender);
+
+            // then we subtract money from the sender's account to pay for gas - this will be paid
+            // even if the entry execution fails
+            _stateProvider.SubtractFromBalance(env.Sender, gasLimit * env.GasPrice, spec);
+
+            // we commit the nonce and gas payment
+            _stateProvider.Commit(_specProvider.GetSpec(env.CurrentBlock.Number), txTracer.IsTracingState ? txTracer : null);
+        }
+
+        /// <summary>
+        /// Refunds are issued as a result of calling specific <see cref="VirtualMachine"/> opcodes like SSTORE or
+        /// SELFDESTRUCT. 
+        /// </summary>
+        /// <param name="gasLimit">Gas limit of the entry needed here because
+        /// (<paramref name="gasLimit"/> - <paramref name="unspentGas"/>) / 2 is a hard cap for the gas refund.</param>
+        /// <param name="unspentGas">Unspent gas of the entry needed here because
+        /// (<paramref name="gasLimit"/> - <paramref name="unspentGas"/>) / 2 is a hard cap for the gas refund.</param>
+        /// <param name="substate">Substate of the transaction before the refunds are issued.</param>
+        /// <param name="env">Details of the execution environment.</param>
+        /// <param name="spec">Provides the refund logic version details.</param>
+        /// <returns>Returns spent gas after the refund.</returns>
+        private ulong Refund(ulong gasLimit,
+            ulong unspentGas,
+            TransactionSubstate substate,
+            ExecutionEnvironment env,
+            IReleaseSpec spec)
+        {
+            ulong spentGas = gasLimit;
+            if (!substate.IsError)
+            {
+                spentGas -= unspentGas;
+                ulong refund = substate.ShouldRevert
+                    ? 0
+                    : Math.Min(spentGas / 2UL, (ulong) (substate.Refund + substate.DestroyList.Count * RefundOf.Destroy));
+
+                if (_logger.IsEnabled(LogEventLevel.Verbose))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)